A33-Fault-Injection.md

```diff
@@ -0,0 +1,203 @@
+ message FaultInjectionPolicy {
+   // The fixed delay before starting the RPC.
+   //
+   // No traffic should be sent until the delay phase ends. By default, the
+   // delay duration is 0 (0 seconds and 0 nanos). If the delay is 0, the RPC
+   // delay feature is disabled and delay_fraction will be ignored.
+   google.protobuf.Duration delay = 1;
+
+   // The probability that this RPC will be delayed.
+   //
+   // The default value is 0. The probability is represented as a fraction over
+   // 1 million (1000000). For example, 0.625% will be 6250, 12.5% will be
+   // 125000. Its value should be clamped into the interval [0, 1000000].
+   uint32 delay_fraction = 2;
+
+   // The RPC status (code and message) for the aborted RPC.
+   //
+   // By default the status code is 0 (OK), which disables the RPC abort
+   // feature and ignores abort_percent. If the control plane only offers
+   // HTTP status code, the application should map it to cardinal RPC status
+   // code.
+   //
+   // By default the status message is empty. If the status message is
+   // specified with non-OK status code, then the status message will be added
+   // to the aborted RPC.
+   google.rpc.Status abort_status = 3;
+
+   // The probability that this RPC will be aborted.
+   //
+   // The default value is 0. The probability is represented as a fraction over
+   // 1 million (1000000). For example, 0.625% will be 6250, 12.5% will be
+   // 125000. Its value should be clamped into the interval [0, 1000000].
+   uint32 abort_fraction = 4;
+ }
+
+ FaultInjectionPolicy fault_injection_policy = 8;
+}
+```
+
+Turned out, this approach requires more engineering effort, but makes the fault injection feature available to non-xDS gRPC applications. gRPC team does consider to improve service config with fault injection settings, and will carry-on this work in a separate proposal in future.
+
+
+### Support for Response Rate Limit Algorithm
+
+Envoy uses the token bucket algorithm to throttle the speed of letting response pass through this filter. It will try to write as much as possible from the buffer, then it will schedule another write when the next token is available. Each token permits to write `(max_kbps * 1024) / SecondDivisor` of bytes. The `SecondDivisor` is a constant 16, which splits 1 second into 16 segments (~63ms apart). Only the body of the HTTP stream is throttled by the response rate limit algorithm. Not headers or trailers.
+
+Envoy implements fault injection is an `HTTPFilter`, which can choose to operate on bytes or HTTP messages or events. It is unclear if gRPC can perform the precise same behavior, since gRPC generally operates on proto messages (only Core has visibility to bytes, not Java/Golang). More importantly, there is not enough consensus for if the timeout timer should start at the beginning of an RPC, or the receipt of the first bytes from peer, or after the entire message is received.
+
+
+### Environment variable protection
```

Comment:
This shouldn't be in the "Alternatives" section.  Please move it up to the "Proposal" section, probably line 120.

Reply:
Updated. Good catch!